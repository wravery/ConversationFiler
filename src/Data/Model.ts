export module Data {
    export interface Message {
        Id: string;
        BodyPreview: string;
        Sender: string;
        ToRecipients: string;
        ParentFolderId: string;
    }

    export interface Folder {
        Id: string;
        DisplayName: string;
    }

    export interface Match {
        message: Message;
        folder: Folder;
    }

    export enum Progress {
        GetCallbackToken,
        GetConversation,
        GetExcludedFolders,
        GetFolderNames,
        Success,
        NotFound,
        Error
    }

    export interface IModel {
        getItemsAsync(onLoadComplete: (results: Match[]) => void, onProgress: (progress: Progress) => void, onError: (message: string) => void): void;
        moveItemsAsync(folderId: string, onMoveComplete: (count: number) => void, onError: (message: string) => void): void;
    }
}