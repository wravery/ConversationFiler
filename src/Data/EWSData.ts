import { Data } from "./Model";

export module EWSData {
    interface ItemId {
        id: string;
        changeKey: string;
    }

    interface FolderData {
        folderId: ItemId;
        displayName?: string;
    }

    interface MessageData {
        itemId: ItemId;
        conversation: ConversationData;
        folder?: FolderData;
        body?: string;
        from?: string;
        to?: string;
    }

    interface ConversationData {
        id: string;
        items: MessageData[];
        global: MessageData[];
    }

    class RequestBuilder {
        private static beginRequest = [
            '<?xml version="1.0" encoding="utf-8" ?>',
            '<soap:Envelope xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"',
            '    xmlns:xsd="http://www.w3.org/2001/XMLSchema"',
            '    xmlns:soap="http://schemas.xmlsoap.org/soap/envelope/"',
            '    xmlns:t="http://schemas.microsoft.com/exchange/services/2006/types"',
            '    xmlns:m="http://schemas.microsoft.com/exchange/services/2006/messages">',
            '<soap:Header>',
            '  <t:RequestServerVersion Version="Exchange2010_SP1" />',
            '</soap:Header>',
            '<soap:Body>',
        ].join('\n');

        private static endRequest = [
            '</soap:Body>',
            '</soap:Envelope>'
        ].join('\n');

        static findConversationRequest = [
            RequestBuilder.beginRequest,
            '  <m:FindConversation>',
            '    <m:IndexedPageItemView BasePoint="Beginning" Offset="0" MaxEntriesReturned="20" />',
            '    <m:ParentFolderId>',
            '      <t:DistinguishedFolderId Id="inbox"/>',
            '    </m:ParentFolderId>',
            '  </m:FindConversation>',
            RequestBuilder.endRequest
        ].join('\n');

        static getItemsRequest(messages: MessageData[]) {
            var builder = [
                RequestBuilder.beginRequest,
                '  <m:GetItem>',
                '    <m:ItemShape>',
                '      <t:BaseShape>IdOnly</t:BaseShape>',
                '      <t:BodyType>Text</t:BodyType>',
                '      <t:AdditionalProperties>',
                '        <t:FieldURI FieldURI="item:ParentFolderId" />',
                '        <t:FieldURI FieldURI="message:Sender" />',
                '        <t:FieldURI FieldURI="message:ToRecipients" />',
                '        <t:FieldURI FieldURI="item:Body" />',
                '      </t:AdditionalProperties>',
                '    </m:ItemShape>',
                '    <m:ItemIds>',
            ];

            messages.map(message => {
                builder.push(`      <t:ItemId Id="${message.itemId.id}" ChangeKey="${message.itemId.changeKey}" />`);
            });

            builder.push(
                '    </m:ItemIds>',
                '  </m:GetItem>',
                RequestBuilder.endRequest);

            return builder.join('\n');
        }

        static excludedFolderIdsRequest = [
            RequestBuilder.beginRequest,
            '  <m:GetFolder>',
            '    <m:FolderShape>',
            '      <t:BaseShape>IdOnly</t:BaseShape>',
            '    </m:FolderShape>',
            '    <m:FolderIds>',
            '      <t:DistinguishedFolderId Id="inbox"/>',
            '      <t:DistinguishedFolderId Id="drafts"/>',
            '      <t:DistinguishedFolderId Id="sentitems"/>',
            '      <t:DistinguishedFolderId Id="deleteditems"/>',
            '    </m:FolderIds>',
            '  </m:GetFolder >',
            RequestBuilder.endRequest
        ].join('\n');

        static getFolderNamesRequest(folders: FolderData[]) {
            var builder = [
                RequestBuilder.beginRequest,
                '  <m:GetFolder>',
                '    <m:FolderShape>',
                '      <t:BaseShape>IdOnly</t:BaseShape>',
                '      <t:AdditionalProperties>',
                '        <t:FieldURI FieldURI="folder:DisplayName" />',
                '      </t:AdditionalProperties>',
                '    </m:FolderShape>',
                '    <m:FolderIds>'
            ];

            folders.map(folder => {
                builder.push(`      <t:FolderId Id="${folder.folderId.id}" ChangeKey="${folder.folderId.changeKey}" />`);
            });

            builder.push(
                '    </m:FolderIds>',
                '  </m:GetFolder >',
                RequestBuilder.endRequest);

            return builder.join('\n');
        }

        static moveItemsRequest(messages: MessageData[], folderId: string) {
            var builder = [
                RequestBuilder.beginRequest,
                '  <m:MoveItem>',
                '    <m:ToFolderId>',
                '      <t:FolderId Id="' + folderId + '"/>',
                '    </m:ToFolderId>',
                '    <m:ItemIds>',
            ];

            messages.map(message => {
                builder.push(`      <t:ItemId Id="${message.itemId.id}" ChangeKey="${message.itemId.changeKey}" />`);
            });

            builder.push(
                '    </m:ItemIds>',
                '  </m:MoveItem>',
                RequestBuilder.endRequest);

            return builder.join('\n');
        }
    }

    class Context {
        constructor(private mailbox: Office.Mailbox) {
            this.itemId = (<Office.ItemRead>this.mailbox.item).itemId;
        }

        private itemId: string;
        private conversationXml?: XMLDocument;
        private conversation?: ConversationData;
        private excludedFolders?: FolderData[];
        private itemsXml?: XMLDocument;
        private folderNamesXml?: XMLDocument;

        private onLoadComplete?: (results: Data.Match[]) => void;
        private onProgress?: (progress: Data.Progress) => void;
        private onError?: (message: string) => void;
        private onMoveComplete?: (count: number) => void

        loadItems(onLoadComplete: (results: Data.Match[]) => void, onProgress: (progress: Data.Progress) => void, onError: (message: string) => void): void {
            this.onLoadComplete = onLoadComplete;
            this.onProgress = onProgress;
            this.onError = onError;

            console.log('Finding the conversation with EWS...');
            this.onProgress(Data.Progress.GetConversation);

            this.mailbox.makeEwsRequestAsync(RequestBuilder.findConversationRequest, (result) => {
                if (!result.value) {
                    this.onError(result.error.message);
                    return;
                }

                this.conversationXml = $.parseXML(result.value);
                this.getConversation();
            });
        }

        private getConversation() {
            const $conversation = $(this.conversationXml.querySelectorAll('Conversation > GlobalItemIds > ItemId'))
                .filter(`[Id="${(<Office.ItemRead>this.mailbox.item).itemId}"]`)
                .parents('Conversation');

            if (!$conversation.length) {
                this.onError("This conversation isn't in your inbox's top 20.");
                return;
            }

            const messageCount = parseInt($conversation.find('MessageCount').text());
            const globalCount = parseInt($conversation.find('GlobalMessageCount').text());

            if (messageCount >= globalCount) {
                this.onLoadComplete([]);
                return;
            }

            let sameFolderItemIds: ItemId[] = [];

            $conversation.find('ItemIds > ItemId').each(function () {
                const $this = $(this);

                sameFolderItemIds.push({
                    id: $this.attr('Id'),
                    changeKey: $this.attr('ChangeKey')
                });
            });

            let otherFolderItemIds: ItemId[] = [];

            $conversation.find('GlobalItemIds > ItemId').each(function () {
                const $this = $(this);

                otherFolderItemIds.push({
                    id: $this.attr('Id'),
                    changeKey: $this.attr('ChangeKey')
                });
            });

            if (!sameFolderItemIds.length || otherFolderItemIds.length <= sameFolderItemIds.length) {
                this.onLoadComplete([]);
                return;
            }

            this.conversation = {
                id: (<Office.Message>this.mailbox.item).conversationId,
                items: sameFolderItemIds.map(itemId => ({ itemId: itemId, conversation: this.conversation })),
                global: otherFolderItemIds.map(itemId => ({ itemId: itemId, conversation: this.conversation }))
            };

            this.loadExcludedFolders();
        }

        private loadExcludedFolders() {
            if (this.excludedFolders) {
                this.loadMessages();
            } else {
                console.log('Getting the list of excluded folders');
                this.onProgress(Data.Progress.GetExcludedFolders);

                this.mailbox.makeEwsRequestAsync(RequestBuilder.excludedFolderIdsRequest, (result) => {
                    if (!result.value) {
                        this.onError(result.error.message);
                        return;
                    }

                    let foldersXml = $.parseXML(result.value);
                    let excludedFolders: FolderData[] = [];

                    $(foldersXml.querySelectorAll('GetFolderResponseMessage > Folders > Folder > FolderId')).each(function () {
                        var $this = $(this);
                        excludedFolders.push({
                            folderId: {
                                id: $this.attr('Id'),
                                changeKey: $this.attr('ChangeKey')
                            }
                        });
                    });

                    this.excludedFolders = excludedFolders;

                    this.loadMessages();
                });
            }
        }

        private loadMessages() {
            if (this.itemsXml) {
                this.getMessages();
            } else {
                console.log(`Getting the messages in other folders: ${this.conversation.global.length}`);
                this.onProgress(Data.Progress.GetConversation);

                this.mailbox.makeEwsRequestAsync(RequestBuilder.getItemsRequest(this.conversation.global), (result) => {
                    if (!result.value) {
                        this.onError(result.error.message);
                        return;
                    }

                    this.itemsXml = $.parseXML(result.value);
                    this.getMessages();
                });
            }
        }

        private getMessages() {
            let $messages = $(this.itemsXml.querySelectorAll('GetItemResponseMessage > Items > Message > ParentFolderId'));

            this.excludedFolders.map(folder => {
                $messages = $messages.filter(`[Id!="${folder.folderId.id}"]`);
            });

            $messages = $messages.parent();

            this.conversation.global.map(item => {
                for (let i = 0; i < $messages.length; i++) {
                    const msg = $messages[i];

                    if (msg.querySelector(`ItemId[Id="${item.itemId.id}"]`)) {
                        let folderId = msg.querySelector('ParentFolderId');

                        item.folder = {
                            folderId: {
                                id: folderId.getAttribute('Id'),
                                changeKey: folderId.getAttribute('ChangeKey')
                            }
                        };
                        item.from = msg.querySelector('Sender > Mailbox > Name').textContent;
                        item.to = msg.querySelector('ToRecipients > Mailbox > Name').textContent;
                        item.body = msg.querySelector('Body').textContent.slice(0, 200);
                        break;
                    }
                }
            });

            this.loadFolderDisplayNames();
        }

        private loadFolderDisplayNames() {
            if (this.folderNamesXml) {
                this.getFolderDisplayNames();
            } else {
                let destinations: FolderData[] = [];

                this.conversation.global.map(item => {
                    if (item.folder) {
                        destinations.push(item.folder);
                    }
                });

                if (!destinations.length) {
                    this.onLoadComplete([]);
                    return;
                }

                console.log(`Getting the display names of the other folders: ${destinations.length}`);
                this.onProgress(Data.Progress.GetFolderNames);

                this.mailbox.makeEwsRequestAsync(RequestBuilder.getFolderNamesRequest(destinations), (result) => {
                    if (!result.value) {
                        this.onError(result.error.message);
                        return;
                    }

                    this.folderNamesXml = $.parseXML(result.value);
                    this.getFolderDisplayNames();
                });
            }
        }

        private getFolderDisplayNames() {
            let matches: Data.Match[] = [];

            this.conversation.global.map((item: MessageData) => {
                if (!item.folder) {
                    return;
                }

                const folder = this.folderNamesXml.querySelector(`GetFolderResponseMessage > Folders > Folder > FolderId[Id="${item.folder.folderId.id}"]`).parentNode;
                item.folder.displayName = (<Element>folder).querySelector('DisplayName').textContent;

                matches.push({
                    message: {
                        Id: item.itemId.id,
                        BodyPreview: item.body,
                        Sender: item.from,
                        ToRecipients: item.to,
                        ParentFolderId: item.folder.folderId.id
                    },
                    folder: {
                        Id: item.folder.folderId.id,
                        DisplayName: item.folder.displayName
                    }
                });
            });

            console.log(`Finished loading items in other folders: ${matches.length}`);
            this.onLoadComplete(Data.removeDuplicates(matches, this.itemId));
        }

        moveItems(folderId: string, onMoveComplete: (count: number) => void, onError: (message: string) => void) {
            this.onMoveComplete = onMoveComplete;
            this.onError = onError;

            console.log(`Moving items to folder: ${folderId}`);

            this.mailbox.makeEwsRequestAsync(RequestBuilder.moveItemsRequest(this.conversation.items, folderId), (result) => {
                if (!result.value) {
                    this.onError(result.error.message);
                    return;
                }

                console.log(`Finished moving items to other folder: ${this.conversation.items.length}`);
                this.onMoveComplete(this.conversation.items.length);
            });
        }
    }

    export class Model implements Data.IModel {
        private context?: Context;

        constructor(mailbox: Office.Mailbox) {
            this.context = new Context(mailbox);
        }

        getItemsAsync(onLoadComplete: (results: Data.Match[]) => void, onProgress: (progress: Data.Progress) => void, onError: (message: string) => void): void {
            this.context.loadItems(onLoadComplete, onProgress, onError);
        }

        moveItemsAsync(folderId: string, onMoveComplete: (count: number) => void, onError: (message: string) => void): void {
            this.context.moveItems(folderId, onMoveComplete, onError);
        }
    }
}