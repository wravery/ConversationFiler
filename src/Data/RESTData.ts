import { Data } from "./Model";

export module RESTData {
    const Endpoint = '/v2.0/me';

    enum ExcludedFolders {
        Inbox,
        Drafts,
        SentItems,
        DeletedItems,

        // Sentinel value for enumerating the folder names
        Count
    }

    interface EmailAddressJson {
        Name: string;
        Address: string;
    }

    interface AddressJson {
        EmailAddress: EmailAddressJson;
    }

    interface MessageJson {
        Id: string;
        BodyPreview: string;
        Sender: AddressJson;
        ToRecipients: AddressJson[];
        ParentFolderId: string;
    }

    interface MessageJsonCollection {
        value: MessageJson[];
    }

    interface FolderJson {
        Id: string;
        DisplayName?: string;
    }

    class Context {
        constructor(private mailbox: Office.Mailbox) {
            this.itemId = this.getRestId((<Office.ItemRead>this.mailbox.item).itemId);
        }

        private itemId: string;
        private token?: string;
        private currentFolderId?: string;
        private conversationMessages?: MessageJson[];
        private excludedFolderIds?: string[];

        private onLoadComplete?: (results: Data.Match[]) => void;
        private onProgress?: (progress: Data.Progress) => void;
        private onError?: (message: string) => void;
        private onMoveComplete?: (count: number) => void

        loadItems(onLoadComplete: (results: Data.Match[]) => void, onProgress: (progress: Data.Progress) => void, onError: (message: string) => void): void {
            this.onLoadComplete = onLoadComplete;
            this.onProgress = onProgress;
            this.onError = onError;

            console.log('Requesting the REST callback token...');
            this.onProgress(Data.Progress.GetCallbackToken);

            // Start the chain of requests by getting a callback token.
            this.mailbox.getCallbackTokenAsync({ isRest: true },
                (result: Office.AsyncResult) => {
                    if (result.status === Office.AsyncResultStatus.Succeeded) {
                        this.getConversation(result);
                    } else {
                        this.onError(result.error.message);
                    }
                });
        }

        // Sometimes we need to make separate REST requests for multiple items. Wait until they all complete and then
        // invoke the callbacks all at once with an array of typed results.
        private collateRequests<T>(requests: JQueryPromise<T>[], onDone: (results: T[]) => void, onFail: (message: string) => void): void {
            if (requests.length > 1) {
                $.when.apply($, requests)
                    .done((...results: any[]) => {
                        onDone(results.map(result => <T>result[0]));
                    }).fail((message: string) => {
                        this.onError(message);
                    });
            } else {
                requests[0]
                    .done((result: T) => {
                        onDone([result]);
                    }).fail((message: string) => {
                        this.onError(message);
                    });
            }
        }

        // If we're on iOS, the IDs we get from Office.context.mailbox.item are already REST IDs. Otherwise we need
        // to convert them from the EWS ID format to the REST ID format.
        private getRestId(itemId: string) {
            if (this.mailbox.diagnostics.hostName === 'OutlookIOS') {
                return itemId;
            }

            return this.mailbox.convertToRestId(itemId, Office.MailboxEnums.RestVersion.v2_0);
        }

        // Send a REST request to retrieve a list of messages in this conversation.
        private getConversation(result: Office.AsyncResult) {
            this.token = <string>result.value;

            const conversationId = (<Office.Message>this.mailbox.item).conversationId;
            const restConversationId = this.getRestId(conversationId);
            const restUrl = `${this.mailbox.restUrl}${Endpoint}/messages?$filter=ConversationId eq '${restConversationId}'&$select=Id,Subject,BodyPreview,Sender,ToRecipients,ParentFolderId`;

            console.log(`Getting the list of items in the conversation: ${restUrl}`);
            this.onProgress(Data.Progress.GetConversation);

            $.ajax({
                url: restUrl,
                async: true,
                dataType: 'json',
                headers: { 'Authorization': `Bearer ${this.token}` }
            }).done((result: MessageJsonCollection) => {
                this.getExcludedFolders(result);
            }).fail((message: string) => {
                this.onError(message);
            });
        }

        // Send a REST request to identify each of the folders we want to exclude in our results.
        private getExcludedFolders(result: MessageJsonCollection) {
            if (!result || !result.value || 0 === result.value.length) {
                this.onLoadComplete([]);
                return;
            }

            this.conversationMessages = result.value;

            // Get the current folderId.
            const currentFolderId = this.conversationMessages
                .filter(value => value.Id === this.itemId)
                .reduce((previousValue: string, value) => value.ParentFolderId, undefined);

            // We should exclude some special folders, but we need to get their folderIds.
            let requests: JQueryXHR[] = [];

            for (let i = 0; i < ExcludedFolders.Count; ++i) {
                const folderId = ExcludedFolders[i];
                const restUrl = `${this.mailbox.restUrl}${Endpoint}/mailfolders/${folderId}?$select=Id`;

                console.log(`Getting excluded folder ID: ${restUrl}`);

                requests.push($.ajax({
                    url: restUrl,
                    async: true,
                    dataType: 'json',
                    headers: { 'Authorization': `Bearer ${this.token}` }
                }));
            }

            this.onProgress(Data.Progress.GetExcludedFolders);

            this.collateRequests(<JQueryPromise<FolderJson>[]>requests, (results) => {
                const excludedFolderIds = results.map(value => value.Id);

                this.getFolderNames(currentFolderId, excludedFolderIds);
            }, (message) => {
                this.onError(message);
            });
        }

        // Send REST requests to fill in the display names of all the folders we are not excluding.
        private getFolderNames(currentFolderId: string, excludedFolderIds: string[]) {
            interface folderMapEntry {
                folder: FolderJson;
                messages: MessageJson[];
            };

            const folderMap = this.conversationMessages
                .filter(message => !excludedFolderIds.reduce((previousValue, value) =>
                    previousValue || value === message.ParentFolderId, false))
                .reduce((previousValue: folderMapEntry[], message) => {
                    const entry = previousValue
                        .filter(value => value.folder.Id === message.ParentFolderId)
                        .pop();

                    if (entry) {
                        // Add this message to the existing entry.
                        entry.messages.push(message);
                    } else {
                        // Create a new entry for this folder.
                        previousValue.push({ folder: { Id: message.ParentFolderId }, messages: [message] });
                    }

                    return previousValue;
                }, []);

            if (folderMap.length === 0) {
                this.onLoadComplete([]);
                return;
            }

            this.currentFolderId = currentFolderId;
            this.excludedFolderIds = excludedFolderIds;

            const requests = folderMap.map((entry) => {
                const restUrl = `${this.mailbox.restUrl}${Endpoint}/mailfolders/${entry.folder.Id}?$select=Id,DisplayName`;

                console.log(`Getting included folder name: ${restUrl}`);

                return $.ajax({
                    url: restUrl,
                    async: true,
                    dataType: 'json',
                    headers: { 'Authorization': `Bearer ${this.token}` }
                });
            });

            this.onProgress(Data.Progress.GetFolderNames);

            this.collateRequests(<JQueryPromise<FolderJson>[]>requests, (results: FolderJson[]) => {
                results.forEach((value) => {
                    for (let i = 0; i < folderMap.length; ++i) {
                        if (folderMap[i].folder.Id === value.Id) {
                            folderMap[i].folder.DisplayName = value.DisplayName;
                            break;
                        }
                    }
                });

                const matches = folderMap.reduce((previousValue: Data.Match[], currentValue) => {
                    previousValue.push.apply(currentValue.messages.map(item => <Data.Match>({
                        message: {
                            Id: item.Id,
                            BodyPreview: item.BodyPreview,
                            Sender: item.Sender.EmailAddress.Name,
                            ToRecipients: item.ToRecipients.map(address => address.EmailAddress.Name).join('; '),
                            ParentFolderId: item.ParentFolderId
                        },
                        folder: {
                            Id: currentValue.folder.Id,
                            DisplayName: currentValue.folder.DisplayName
                        }
                    })));

                    return previousValue;
                }, []);

                console.log(`Finished loading items in other folders: ${matches.length}`);
                this.onLoadComplete(Data.removeDuplicates(matches, this.itemId));
            }, (message: string) => {
                this.onError(message);
            });
        }

        moveItems(folderId: string, onMoveComplete: (count: number) => void, onError: (message: string) => void) {
            this.onMoveComplete = onMoveComplete;
            this.onError = onError;

            console.log(`Moving items to folder: ${folderId}`);

            const requests = this.conversationMessages
                .filter(message => message.ParentFolderId !== this.currentFolderId)
                .map(message => {
                    const restUrl = `${this.mailbox.restUrl}${Endpoint}/messages/${message.Id}/move`;

                    console.log(`Moving item: ${restUrl}`);

                    return $.ajax({
                        url: restUrl,
                        async: true,
                        method: 'POST',
                        contentType: 'application/json',
                        dataType: 'json',
                        data: JSON.stringify({ DestinationId: folderId }),
                        headers: { 'Authorization': `Bearer ${this.token}` }
                    })
                });

            this.collateRequests(<JQueryPromise<MessageJson>[]>requests, (results: MessageJson[]) => {
                console.log(`Finished moving items to other folder: ${results.length}`);
                this.onMoveComplete(results.length);
            }, (message: string) => {
                this.onError(message);
            });
        }
    }

    export class Model implements Data.IModel {
        private context?: Context;

        constructor(mailbox: Office.Mailbox) {
            this.context = new Context(mailbox);
        }

        getItemsAsync(onLoadComplete: (results: Data.Match[]) => void, onProgress: (progress: Data.Progress) => void, onError: (message: string) => void): void {
            this.context.loadItems(onLoadComplete, onProgress, onError);
        }

        moveItemsAsync(folderId: string, onMoveComplete: (count: number) => void, onError: (message: string) => void): void {
            this.context.moveItems(folderId, onMoveComplete, onError);
        }
    }
}