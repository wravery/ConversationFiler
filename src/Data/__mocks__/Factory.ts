import { Data } from "../Model";

export module Factory {
    export interface MockMailbox extends Office.Mailbox {
        mockResults: Data.Match[];
    }

    class MockData implements Data.IModel {
        constructor(private mockResults: Data.Match[]) {
        }

        getItemsAsync(onLoadComplete: (results: Data.Match[]) => void, onProgress: (progress: Data.Progress) => void, onError: (message: string) => void): void {
            if (this.mockResults) {
                if (this.mockResults.length > 0) {
                    onLoadComplete(this.mockResults);
                } else {
                    onError("Empty results");
                }
            } else {
                onProgress(Data.Progress.GetCallbackToken);
            }
        }

        moveItemsAsync(folderId: string, onMoveComplete: (count: number) => void, onError: (message: string) => void): void {
            onMoveComplete(this.mockResults.length);
        }
    }

    // Use the MockData provider
    export function getData(mailbox: Office.Mailbox): Data.IModel {
        return new MockData((<MockMailbox>mailbox).mockResults);
    }
}