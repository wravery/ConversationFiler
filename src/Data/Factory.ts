import { Data } from "./Model";
import { RESTData } from "./RESTData";
import { EWSData } from "./EWSData";

export module Factory {
    // Use the RESTData provider if possible, but if it's not supported, fallback to the EWSData provider
    export function getData(mailbox: Office.Mailbox): Data.IModel {
        return mailbox.restUrl
            ? new RESTData.Model(mailbox)
            : new EWSData.Model(mailbox);
    }
}