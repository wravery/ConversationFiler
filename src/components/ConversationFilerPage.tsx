import * as React from "react";
import * as JQuery from "jquery";

import { Data } from "../Data/Model"
import { Factory } from "../Data/Factory";

import { StatusMessage } from "./StatusMessage";
import { SearchResults } from "./SearchResults";
import { Feedback } from "./Feedback";

export interface ConversationFilerPageProps {
    mailbox: Office.Mailbox;
}

interface ConversationFilerPageState {
    progress: Data.Progress;
    data?: Data.IModel;
    error?: string;
    matches?: Data.Match[];
}

export class ConversationFilerPage extends React.Component<ConversationFilerPageProps, ConversationFilerPageState> {
    constructor(props: ConversationFilerPageProps) {
        super(props);
        this.onSelection = this.onSelection.bind(this);

        this.state = { progress: Data.Progress.GetCallbackToken };
    }

    // Start the chain of requests to load the data
    componentDidMount() {
        const data = Factory.getData(this.props.mailbox);
 
        this.setState({ data: data });

        console.log('Starting to load the conversation...');

        data.getItemsAsync((results) => {
            console.log(`Loaded the conversation: ${results.length}`);

            if (results.length > 0) {
                this.setState({ progress: Data.Progress.Success, matches: results });
            } else {
                this.setState({ progress: Data.Progress.NotFound });
            }
        }, (progress) => {
            console.log(`Progress loading the conversation: ${Data.Progress[progress]}`);

            this.setState({ progress: progress });
        }, (message) => {
            console.log(`Error loading the conversation: ${message}`);

            this.setState({ progress: Data.Progress.Error, error: message });
        });
    }

    private onSelection(folderId: string) {
        console.log(`Selected a folder: ${folderId}`);

        this.state.data.moveItemsAsync(folderId, (count) => {
            console.log(`Finished moving the items: ${count}`);
        }, (message) => {
            this.setState({ progress: Data.Progress.Error, error: message });
        });
    }

    render() {
        return (<div>
            <StatusMessage progress={this.state.progress} message={this.state.error} />
            <SearchResults matches={this.state.matches} onSelection={this.onSelection} />
            <Feedback />
        </div>);
    }
}
