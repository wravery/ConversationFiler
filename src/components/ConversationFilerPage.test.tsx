import * as React from "react";
import * as ReactDOM from "react-dom";
import * as TestUtils from "react-addons-test-utils";

import { Data } from "../Data/Model";
import { Factory } from "../Data/__mocks__/Factory";
jest.mock('../Data/Factory');

import { ConversationFilerPage } from "./ConversationFilerPage";

test("Loading", () => {
    const mockMailbox = {
        restUrl: 'https://foo.bar.com/api'
    } as Factory.MockMailbox;

    const log = console.log;
    console.log = jest.fn((message: string) => {
        expect(message).toMatchSnapshot();
    });

    const component = TestUtils.renderIntoDocument(<ConversationFilerPage mailbox={mockMailbox} />) as ConversationFilerPage;
    const rendered = ReactDOM.findDOMNode(component);
    expect(rendered.innerHTML).toMatchSnapshot();

    console.log = log;
});

test("Empty results", () => {
    const mockMailbox = {
        restUrl: 'https://foo.bar.com/api',
        mockResults: []
    } as Factory.MockMailbox;

    const log = console.log;
    console.log = jest.fn((message: string) => {
        expect(message).toMatchSnapshot();
    });

    const component = TestUtils.renderIntoDocument(<ConversationFilerPage mailbox={mockMailbox} />) as ConversationFilerPage;
    const rendered = ReactDOM.findDOMNode(component);
    expect(rendered.innerHTML).toMatchSnapshot();

    console.log = log;
});

test("Dummy data", () => {
    const mockMailbox = {
        restUrl: 'https://foo.bar.com/api',
        mockResults: [{
            folder: {
                Id: 'folderId1',
                DisplayName: 'Folder 1'
            },
            message: {
                Id: 'messageId1',
                BodyPreview: 'Here\'s a preview of a message body',
                Sender: 'Foo Bar',
                ToRecipients: 'Baz Bar',
                ParentFolderId: 'folderId1'
            }
        }, {
            folder: {
                Id: 'folderId2',
                DisplayName: 'Folder 2'
            },
            message: {
                Id: 'messageId2',
                BodyPreview: 'Here\'s another message body',
                Sender: 'Baz Bar',
                ToRecipients: 'Foo Bar',
                ParentFolderId: 'folderId2'
            }
        }]
    } as Factory.MockMailbox;

    const log = console.log;
    console.log = jest.fn((message: string) => {
        expect(message).toMatchSnapshot();
    });

    const component = TestUtils.renderIntoDocument(<ConversationFilerPage mailbox={mockMailbox} />) as ConversationFilerPage;
    const rendered = ReactDOM.findDOMNode(component);
    expect(rendered.innerHTML).toMatchSnapshot();

    console.log = log;
});
