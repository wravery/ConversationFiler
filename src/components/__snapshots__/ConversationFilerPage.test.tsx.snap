// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dummy data 1`] = `"Starting to load the conversation..."`;

exports[`Dummy data 2`] = `"Loaded the conversation: 2"`;

exports[`Dummy data 3`] = `"<!-- react-empty: 8 --><div><span class=\\"ms-font-l\\">I found some items in this conversation filed in other folders. Click on one of the folders listed here to automatically reunite this part of the conversation with the ones that came before:</span><br><div class=\\"ms-Viewport\\" style=\\"min-width: 1px; min-height: 1px;\\"></div></div><div class=\\"feedback ms-font-s\\"><a href=\\"https://beandotnet.azurewebsites.net/\\">about this app</a><!-- react-text: 6 -->&nbsp;<!-- /react-text --><a href=\\"mailto:wravery@hotmail.com?Subject=Conversation%20Filer%20v2.0%20App%20for%20Outlook\\">send feedback</a></div>"`;

exports[`Empty results 1`] = `"Starting to load the conversation..."`;

exports[`Empty results 2`] = `"Error loading the conversation: Empty results"`;

exports[`Empty results 3`] = `"<div><span class=\\"ms-font-l\\">Sorry, I couldn't figure out where this message should go. :(</span><br><span class=\\"ms-font-m\\">Empty results</span></div><!-- react-empty: 3 --><div class=\\"feedback ms-font-s\\"><a href=\\"https://beandotnet.azurewebsites.net/\\">about this app</a><!-- react-text: 6 -->&nbsp;<!-- /react-text --><a href=\\"mailto:wravery@hotmail.com?Subject=Conversation%20Filer%20v2.0%20App%20for%20Outlook\\">send feedback</a></div>"`;

exports[`Loading 1`] = `"Starting to load the conversation..."`;

exports[`Loading 2`] = `"Progress loading the conversation: GetCallbackToken"`;

exports[`Loading 3`] = `"<span class=\\"ms-font-l\\">Looking for other messages in this conversation...</span><!-- react-empty: 3 --><div class=\\"feedback ms-font-s\\"><a href=\\"https://beandotnet.azurewebsites.net/\\">about this app</a><!-- react-text: 6 -->&nbsp;<!-- /react-text --><a href=\\"mailto:wravery@hotmail.com?Subject=Conversation%20Filer%20v2.0%20App%20for%20Outlook\\">send feedback</a></div>"`;
